text/0013-prepare-blockbuilder-and-core-runtime-apis-for-mbms.md

```diff
@@ -0,0 +1,127 @@
+| **Start Date**  | July 24, 2023 |
+| **Description** | Prepare the `BlockBuilder` and `Core` Runtime APIs for Multi-Block-Migrations.|
+| **Authors**     | Oliver Tale-Yazdi |
+
+## Summary
+
+Introduces breaking changes to the `BlockBuilder` and `Core` runtime APIs and bumps them to version
+5 and 7 respectively. A new function `after_inherents` is added to the `BlockBuilder` runtime API
+and the `initialize_block` function of `Core` is changed to return an enum to indicate what
+extrinsics can be applied.
+
+## Motivation
+
+The original motivation is Multi-Block-Migrations. They require the runtime to be able to tell the
+block builder that it should not attempt to include transactions in the current block. Currently,
+there is no communication possible between runtime and block builder that could achieve this.
+Further, it is necessary to execute runtime logic right after inherent application but still before
+transaction inclusion.
+
+This RFC proposes a way of communication between the runtime and the block builder by changing the
+`initialize_block` function to return a `ExtrinsicInclusionMode` enum. Additionally, an
+`after_inherents` function is introduced that runs after inherent but before transaction
+application.  
+
+## Stakeholders
+
+- Substrate Maintainers: They will have to implement this upstream with all the testing, audit and
+  maintenance burden.
+- Polkadot Runtime developers: They will have to adapt to this breaking change.
+- Polkadot Parachain Teams: They also have to adapt to the breaking changes but then eventually have
+  multi-block migrations available.
+
+## Explanation
+
+The only relevant change is on the node side in the block authoring logic. Any further preparation
+for MBMs can happen entirely on the runtime side with the provided primitives.
+
+All block authors MUST respect the `ExtrinsicInclusionMode` that is returned by `initialize_block`.
+They MUST always invoke `after_inherents` directly after inherent application. The runtime MAY
+reject a block that violates either of those requirements.
+
+Enum `ExtrinsicInclusionMode` has two variants:  
+- `AllExtrinsics`: All extrinsics can be applied. It is the default behaviour prior to this RFC.
+- `OnlyInherents`: Only inherents SHALL be applied by the block author. This differs from the
+  current behaviour by omitting transactions.
+
+It is RECOMMENDED that block authors keep transactions in the local transaction pool (if applicable)
+for as long as `initialize_block` returns `OnlyInherents`.  
+Backwards compatibility with the current runtime API SHOULD be implemented on the node side to not
+mandate a lockstep update.
+
+## Drawbacks
+
+[…] drawbacks relating to performance, ergonomics, user experience, security, or privacy: None
+
+The only drawback is that the block execution logic becomes more complicated. There is more room for
+error.  
+Downstream developers will also need to adapt their code to this breaking change.
```

Comment:
Do we have a good definition of breaking changes? This seems to be opt-in and doesn't break node-side code unless the user's runtime is updated.